### DISCUSSION TITLE
how to hand the json

### DISCUSSION THREAD

**Discussion by elueckel** [Art-of-WiFi/UniFi-API-client#124] [Q&A] [answered]
Hi,
me - again ... the guy with the token (which is working by now). While this is an indirect question, maybe you guys can help me. 

As mentioned I am working on a much smaller code than this full library is and where I am stuck is to block a device. 

I am encoding the device into the a JSON via 
`$CommandToController = json_encode(array(
					"cmd" => $Command,
					"mac" => $DeviceMacAdress));`

Where obviously I am using $Command = "block-sta";

I am doing a `curl_setopt($ch, CURLOPT_URL, "https://192.168.1.1/proxy/network/api/s/default/cmd/stamgr/");`

To be followed by
`curl_setopt($ch, CURLOPT_HTTPHEADER, array("cookie: ".$Cookie,'Content-Type: application/json',"Expect:"));	
curl_setopt($ch, CURLOPT_POSTFIELDS, "json=".$CommandToController);
curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, FALSE); 
curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, FALSE);	
curl_setopt($ch, CURLOPT_SSLVERSION, 'CURL_SSLVERSION_TLSv1');
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true); 	
$RawData = curl_exec($ch);`

and always end up with the error  
`22.08.2021, 19:24:15 |       Device Blocker | Feedback from UniFi Controller: Not Found / HTTP Message 404`

It seems I am too stupid to whether compose the URL or encode the JSON object. 

I have spent hours looking at the code you guys build and I am not able to get mine working. 

Again - fully understood if you cannot help me here, but I am stuck on this topic ... when I cal the URL in the browser, I am also getting a {"meta":{"rc":"error","msg":"api.err.NotFound"},"data":[]}

Thanks

**Comment by thib3113** [comment]
and without the trailing slash ?

**Reply by elueckel** [reply]
Behind the stamgr ? Will try.

**Reply by elueckel** [reply]
Removing the / does not make a difference.

**Reply by elueckel** [reply]
I have a suspicion that it has something to do with the way the JSON with Command : Max is pass on. 

Funny enough I have tried it with something different like "evtmgt / archive-all-alarms" and this resulted in the same error. Also when I am encoding the JSON into a URL, this is also not accepted - I tried to make sure, that I am properly authenticated.

**Reply by thib3113** [reply]
Ok, sorry ... 

I doesn't understand all you're code ( a long time I didn't use PHP to developp )  :/ ... But maybe a reformating can help . 

Hum, as I remember, you need to set CURLOPT_POST to do a post ? ( or the CURLOPT_POSTFIELDS is enough ? )

And are you sure about `CURLOPT_POSTFIELDS` `json=XXX` ? ( some tutorials seems to send directly the json, and not json= )

**Reply by elueckel** [reply]
Thanks again - I have the CURLOPT_POST, True in the code, but did not post it in the message.

The json= ... I saw some examples where it was used, so I have tried with it, but also without - it does not make a difference.

**Reply by thib3113** [reply]
Ok, sorry for the delay, when I send a block-sta, here is the curl request I do : 

```
curl -X POST -H "Content-Type:application/json" -H "Accept:application/json" -H "X-CSRF-Token:<csrf_token>" -H "Cookie:TOKEN=myToken" --data '{"cmd":"block-sta","mac":"04:18:d6:85:ed:f2"}' "https://unifi/proxy/network/api/s/default/cmd/stamgr"
```
And it's work ... 

I don't know if it's help

**Comment by elueckel** [comment, marked as answer]
I GOT IT WORKING :-)

The thing which had been missing was the XSRF Token handling. If this is not around, the whole thing does not work. I think I have not had a single module consuming so much time to understand the APIs etc.

Really impressed how the Unifi Client was developed by the Art of WIFI - without it, NO way I would have been able to understand how it all works.

**Reply by malle-pietje** [reply]
Yes, this was a huge puzzle when the UDM (PRO)s were introduced. Also to make the API client class handles the different methods automatically took quite some effort. Glad it helped you create your own solution!

**Reply by thib3113** [reply]
@malle-pietje only "when the UDMs" comes ? not the same with cloud keys ?

**Comment by malle-pietje** [comment]
CloudKey firmware updates with these changes came a long time after the UDMs were introduced.
